CRM/Contribute/Tokens.php

```diff
@@ -73,35 +67,58 @@ protected function getExposedFields(): array {
       'thankyou_date',
       'tax_amount',
       'contribution_status_id',
+      'contribution_status_id:name',
+      'contribution_status_id:label',
       'financial_type_id',
+      'financial_type_id:name',
+      'financial_type_id:label',
       'payment_instrument_id',
+      'payment_instrument_id:name',
+      'payment_instrument_id:label',
       'cancel_reason',
       'amount_level',
       'check_number',
     ];
     if (CRM_Campaign_BAO_Campaign::isCampaignEnable()) {
       $fields[] = 'campaign_id';
+      $fields[] = 'campaign_id.name';
+      $fields[] = 'campaign_id.title';
     }
+
     return $fields;
   }
 
-  /**
-   * Get tokens supporting the syntax we are migrating to.
-   *
-   * In general these are tokens that were not previously supported
-   * so we can add them in the preferred way or that we have
-   * undertaken some, as yet to be written, db update.
-   *
-   * See https://lab.civicrm.org/dev/core/-/issues/2650
-   *
-   * @return string[]
-   */
-  public function getBasicTokens(): array {
-    $return = [];
-    foreach ($this->getExposedFields() as $fieldName) {
-      $return[$fieldName] = $this->getFieldMetadata()[$fieldName]['title'];
+  public function getAliasTokens(): array {
+    $aliases = [];
+    if (CRM_Campaign_BAO_Campaign::isCampaignEnable()) {
```

Comment:
none of this is required now - once mine is merged you just need to make https://github.com/civicrm/civicrm-core/pull/21137 pass

Reply:
I don't know what you mean by that. #21137 seems irrelevant.

This code ensures continued support for the new token `{contribution.campaign_id:label}` -- which you had locked in via unit-test in anticipation of APIv4 support -- but which hasn't been supported by APIv4. Something has to bridge the gap.

When/if APIv4 gets an update to support `{contribution.campaign_id:label}`, *then* this becomes unnecessary - and a tiny patch can remove it. (Just drop the line that aliases `campaign_id:label` to `campaign_id.title` - it won't be hurtful; it'll just be superfluous.) But from my POV, APIv4 support for `campaign_id:label` is a separate rabbit hole. I don't know if that'll be fixed in 5 minutes or 5 weeks, and I don't think it should matter.

This approach keeps the sequence of PRs flexible. The outcome is functionally correct regardless of the order in which the changes are made.

Reply:
@totten we fixed it to be supported by apiv - that was the conversation we had last week

Reply:
But with https://github.com/civicrm/civicrm-core/pull/21109 passed the issue 2745 is met and the pseudotokens are handled in a generic way (including campaign id) - and we wind up with a list of opt out tokens rather than opt in - since we agreed that we wanted to handle the pseudoconstants (including campaign_id) in a consistent way we don't need to list them out - we can determine them from the metadata